doc/codegen/modular_codegen_design.md

```diff
@@ -0,0 +1,571 @@
+- **`control_flow.py`**: Structured control flow extraction
+- **`compiler.py`**: CMake integration and compilation
+
+#### Target System:
+- **`targets/target.py`**: Base target interface
+- **`targets/cpu.py`**: CPU/OpenMP code generation
+- **`targets/cuda.py`**: CUDA/HIP GPU code generation
+- **`targets/fpga.py`**: FPGA code generation
+- **`targets/cpp.py`**: C++ utilities
+- **`targets/mpi.py`**: MPI parallelization
+- **`targets/rtl.py`**: RTL/SystemVerilog generation
+
+#### Specialized Systems:
+- **`targets/sve/`**: ARM SVE vectorization
+- **`targets/mlir/`**: MLIR backend
+- **`instrumentation/`**: Profiling and monitoring
+- **`tools/`**: Type inference and runtime utilities
+
+### Current Monolithic Behaviors in DaCeCodeGenerator
+
+The `DaCeCodeGenerator` class currently handles numerous responsibilities that should be decomposed:
+
+1. **Metadata Collection**
+   - Free symbol analysis (`free_symbols()`)
+   - Symbol and constant resolution
+   - Argument list construction
+   - Shared transient identification
+
+2. **Allocation Analysis**
+   - Allocation lifetime determination (`determine_allocation_lifetime()`)
+   - Allocation scope determination
+   - Memory layout analysis
+   - Persistent/global memory handling
+
+3. **Code Structure Creation**
+   - State struct generation
+   - Function signature creation
+   - Header/footer generation
+   - Constants definition
+
+4. **Traversal and Generation**
+   - Recursive SDFG traversal
+   - State code generation dispatch
+   - Target-specific code routing
+   - Memory allocation/deallocation code
+
+## Proposed Pass Decomposition
+
+### Phase 1: Analysis Passes
+
+#### 1. **TypeInferencePass**
+- **Purpose**: Infer connector types and set default storage/schedule types
+- **Input**: Raw SDFG
+- **Output**: SDFG with inferred types, pipeline_results["type_info"]
+- **Current Location**: `infer_types.py` functions
+
+#### 2. **LibraryExpansionPass**
+- **Purpose**: Expand all library nodes that haven't been expanded
+- **Input**: Type-inferred SDFG
+- **Output**: SDFG with expanded library nodes
+- **Current Location**: `sdfg.expand_library_nodes()`
+
+#### 3. **MetadataCollectionPass**
+- **Purpose**: Collect free symbols, argument lists, constants, shared transients
+- **Input**: Expanded SDFG
+- **Output**: pipeline_results["metadata"] = {symbols, arglist, constants, shared_transients}
+- **Current Location**: `DaCeCodeGenerator.__init__()`
+
+#### 4. **AllocationAnalysisPass**
+- **Purpose**: Determine allocation lifetimes and scopes for all data containers
+- **Input**: SDFG with metadata
+- **Output**: pipeline_results["allocation_info"] = {to_allocate, where_allocated}
+- **Current Location**: `DaCeCodeGenerator.determine_allocation_lifetime()`
+
+#### 5. **ControlFlowAnalysisPass**
+- **Purpose**: Extract structured control flow from state machines
+- **Input**: SDFG
+- **Output**: pipeline_results["control_flow_tree"]
```

Comment:
Even if we leave control flow detection in as part of the pipeline, the output should no longer be a "control flow tree", but rather an SDFG with explicit control flow regions and any unstructured control flow isolated in corresponding control flow regions (as per PR #2045). In that case - still, assuming we even want control flow detection to be part of the code generation pipeline - this should be performed as one of the very first steps of the pipeline, before type inference etc. and most likely right behind Validation, or even before that. My reasoning behind that is that it is significantly easier to perform each individual step, if the assumption can be made that only explicit control flow is present in the SDFG, and state machines are kept to an absolute minimum.

Reply:
Done.